File: solutions/blind-75/array/productOfArrayExceptSelf.js
const productExceptSelf = (nums) => {
    const result = [];
    
    let productSoFar = 1;
    for (let i = 0; i < nums.length; i++) {
        result[i] = productSoFar;
        productSoFar *= nums[i];
    }
    
    productSoFar = 1
    for (let i = nums.length - 1; i >= 0; i--) {
        result[i] *= productSoFar;
        productSoFar *= nums[i];
    }
    
    return result;
};

File: solutions/python/general-algos/arrays/product_of_array_except_self.py

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        """
        First find products from beginning to index. Then find products from end to index. Multiply these together to get products except self.
        """
        result = []

        product_so_far = 1
        for i in range(len(nums)):
            result.append(product_so_far)
            product_so_far *= nums[i]

        product_so_far = 1
        for i in range(len(nums) - 1, -1, -1):
            result[i] *= product_so_far
            product_so_far *= nums[i]

        return result